File: app/models/citacoes_bm.py
"""Modelos do módulo Citações BM (Banco Master).

Monitora processos do escritório "Banco Master / Réu" via DataJud (API
pública do CNJ) para detectar a CITAÇÃO efetiva — gatilho da habilitação.
O sistema só TRAZ as movimentações; quem decide se houve citação é o
operador (status_citacao é alterado exclusivamente por ele).

Tabelas (prefixo cit*):
- cit_processos  — processo monitorado (1 por CNJ).
- cit_movimentos — movimentações capturadas do DataJud (N por processo).
"""

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.session import Base

# Status do processo no fluxo de citação. Alterado EXCLUSIVAMENTE pelo
# operador (o scan automático nunca mexe nisto).
STATUS_PENDENTE = "PENDENTE"
STATUS_CITADO = "CITADO"
STATUS_NAO_CITADO = "NAO_CITADO"

# De onde o processo entrou no monitoramento.
ORIGEM_LISTA = "LISTA_MANUAL"
ORIGEM_L1_AUTO = "L1_AUTO"

# Resultado da última varredura no DataJud.
SCAN_OK = "OK"
SCAN_SEM_HITS = "SEM_HITS"  # DataJud ainda não tem o processo (lag do tribunal)
SCAN_ERRO = "ERRO"


class CitacaoBMProcesso(Base):
    __tablename__ = "cit_processos"

    id = Column(Integer, primary_key=True, index=True)

    # CNJ só com dígitos (20) é a chave canônica; máscara é só pra exibição.
    cnj = Column(String(25), nullable=False, unique=True, index=True)
    cnj_mask = Column(String(40), nullable=True)

    # Resolução no Legal One (pode faltar se o processo ainda não tem pasta).
    lawsuit_id = Column(Integer, nullable=True, index=True)
    office_external_id = Column(Integer, nullable=True, index=True)
    office_path = Column(String, nullable=True)
    l1_creation_date = Column(DateTime(timezone=True), nullable=True)

    # Roteamento DataJud + metadados de exibição.
    tribunal_alias = Column(String(40), nullable=True)
    uf = Column(String(4), nullable=True)
    cidade = Column(String, nullable=True)
    acao = Column(String, nullable=True)
    cliente = Column(String, nullable=True)
    contrario = Column(String, nullable=True)

    origem = Column(String(16), nullable=False, default=ORIGEM_LISTA)

    # Decisão do operador.
    status_citacao = Column(
        String(16), nullable=False, default=STATUS_PENDENTE, index=True
    )
    citado_por_user_id = Column(Integer, nullable=True)
    citado_por_nome = Column(String, nullable=True)
    citado_em = Column(DateTime(timezone=True), nullable=True)
    observacao = Column(Text, nullable=True)

    # Vira False quando o operador marca CITADO (arquiva → sai da varredura).
    monitoramento_ativo = Column(
        Boolean, nullable=False, default=True, index=True
    )

    # Estado da varredura DataJud + contadores denormalizados (lista rápida).
    last_scan_at = Column(DateTime(timezone=True), nullable=True)
    last_scan_status = Column(String(16), nullable=True)
    last_scan_error = Column(Text, nullable=True)
    last_movement_at = Column(DateTime(timezone=True), nullable=True)
    total_movimentos = Column(Integer, nullable=False, default=0)
    novos_movimentos = Column(Integer, nullable=False, default=0)
    tem_candidato_citacao = Column(Boolean, nullable=False, default=False)

    created_by_email = Column(String, nullable=True)
    created_at = Column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), nullable=True)

    movimentos = relationship(
        "CitacaoBMMovimento",
        back_populates="processo",
        cascade="all, delete-orphan",
        order_by="CitacaoBMMovimento.data_hora.desc()",
    )


class CitacaoBMMovimento(Base):
    __tablename__ = "cit_movimentos"
    __table_args__ = (
        # Dedupe: cada fingerprint é único dentro do processo.
        UniqueConstraint("processo_id", "fingerprint", name="uq_cit_mov_fp"),
    )

    id = Column(Integer, primary_key=True, index=True)
    processo_id = Column(
        Integer,
        ForeignKey("cit_processos.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    codigo_tpu = Column(Integer, nullable=True)
    nome = Column(String, nullable=False)
    grau = Column(String(8), nullable=True)
    data_hora = Column(DateTime(timezone=True), nullable=True, index=True)
    complementos = Column(JSONB, nullable=True)
    orgao_julgador = Column(String, nullable=True)

    # Hash estável (grau+codigo+data+nome+complementos) pra evitar reinserção.
    fingerprint = Column(String(64), nullable=False, index=True)

    # Heurística de citação — só DESTACA; quem decide é o operador.
    is_candidato_citacao = Column(
        Boolean, nullable=False, default=False, index=True
    )
    cit_match_termo = Column(String, nullable=True)

    # Não-lido = movimento novo desde a última visita do operador.
    lido = Column(Boolean, nullable=False, default=False, index=True)
    captured_at = Column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    processo = relationship("CitacaoBMProcesso", back_populates="movimentos")

File: frontend/src/services/citacoesBm.ts
// Serviço do módulo Citações BM (monitoramento de citação via DataJud).
// Self-contained (tipos + chamadas) pra não inflar o api.ts gigante.

import { apiFetch } from "@/lib/api-client";

const BASE = "/api/v1/publications/citacoes-bm";

export type StatusCitacao = "PENDENTE" | "CITADO" | "NAO_CITADO";
export type OrigemCitacao = "LISTA_MANUAL" | "L1_AUTO";

export interface CitacaoBMProcesso {
  id: number;
  cnj: string;
  cnj_mask: string | null;
  lawsuit_id: number | null;
  l1_url: string | null;
  office_external_id: number | null;
  office_path: string | null;
  l1_creation_date: string | null;
  tribunal_alias: string | null;
  uf: string | null;
  cidade: string | null;
  acao: string | null;
  cliente: string | null;
  contrario: string | null;
  origem: OrigemCitacao;
  status_citacao: StatusCitacao;
  citado_por_nome: string | null;
  citado_em: string | null;
  observacao: string | null;
  monitoramento_ativo: boolean;
  last_scan_at: string | null;
  last_scan_status: string | null;
  last_movement_at: string | null;
  total_movimentos: number;
  novos_movimentos: number;
  tem_candidato_citacao: boolean;
  created_at: string | null;
}

export interface CitacaoBMMovimento {
  id: number;
  codigo_tpu: number | null;
  nome: string;
  grau: string | null;
  data_hora: string | null;
  complementos: Array<Record<string, unknown>> | null;
  orgao_julgador: string | null;
  is_candidato_citacao: boolean;
  cit_match_termo: string | null;
  lido: boolean;
  captured_at: string | null;
}

export interface CitacaoBMDetail extends CitacaoBMProcesso {
  movimentos: CitacaoBMMovimento[];
  candidatos_count: number;
}

export interface CitacaoBMListResponse {
  total: number;
  items: CitacaoBMProcesso[];
}

export interface CitacaoBMSummary {
  total: number;
  monitorando: number;
  pendentes: number;
  com_novos: number;
  com_candidato: number;
  citados: number;
  nao_citados: number;
}

export interface ListParams {
  status?: StatusCitacao | "";
  origem?: OrigemCitacao | "";
  tribunal_alias?: string;
  uf?: string;
  apenas_com_novos?: boolean;
  arquivados?: "ativos" | "arquivados" | "todos";
  q?: string;
  limit?: number;
  offset?: number;
}

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let detail = `HTTP ${res.status}`;
    try {
      const body = await res.json();
      detail = body?.detail || detail;
    } catch {
      /* ignore */
    }
    throw new Error(detail);
  }
  return res.json() as Promise<T>;
}

export async function getCitacaoSummary(): Promise<CitacaoBMSummary> {
  return json(await apiFetch(`${BASE}/summary`));
}

export async function listCitacaoProcessos(
  params: ListParams,
): Promise<CitacaoBMListResponse> {
  const qs = new URLSearchParams();
  if (params.status) qs.set("status", params.status);
  if (params.origem) qs.set("origem", params.origem);
  if (params.tribunal_alias) qs.set("tribunal_alias", params.tribunal_alias);
  if (params.uf) qs.set("uf", params.uf);
  if (params.apenas_com_novos) qs.set("apenas_com_novos", "true");
  if (params.arquivados) qs.set("arquivados", params.arquivados);
  if (params.q) qs.set("q", params.q);
  qs.set("limit", String(params.limit ?? 50));
  qs.set("offset", String(params.offset ?? 0));
  return json(await apiFetch(`${BASE}?${qs.toString()}`));
}

export async function getCitacaoProcesso(id: number): Promise<CitacaoBMDetail> {
  return json(await apiFetch(`${BASE}/${id}`));
}

export async function ingestCitacaoList(
  text: string,
): Promise<Record<string, unknown>> {
  return json(
    await apiFetch(`${BASE}/ingest/list`, {
      method: "POST",
      body: JSON.stringify({ text }),
    }),
  );
}

export async function ingestCitacaoL1(
  data_corte?: string,
): Promise<Record<string, unknown>> {
  return json(
    await apiFetch(`${BASE}/ingest/l1`, {
      method: "POST",
      body: JSON.stringify({ data_corte: data_corte || null }),
    }),
  );
}

export async function scanCitacaoAll(
  limit?: number,
): Promise<{ started: boolean }> {
  return json(
    await apiFetch(`${BASE}/scan`, {
      method: "POST",
      body: JSON.stringify({ limit: limit ?? null }),
    }),
  );
}

export async function scanCitacaoProcesso(
  id: number,
): Promise<{ status: string; novos: number }> {
  return json(await apiFetch(`${BASE}/${id}/scan`, { method: "POST" }));
}

export async function setCitacaoStatus(
  id: number,
  status: StatusCitacao,
  observacao?: string,
): Promise<CitacaoBMProcesso> {
  return json(
    await apiFetch(`${BASE}/${id}/citacao`, {
      method: "PATCH",
      body: JSON.stringify({ status, observacao: observacao ?? null }),
    }),
  );
}

export async function markCitacaoRead(
  id: number,
): Promise<{ marcados: number }> {
  return json(await apiFetch(`${BASE}/${id}/mark-read`, { method: "POST" }));
}
